runtime/hsa-runtime/core/runtime/amd_aie_aql_queue.cpp

```diff
@@ -193,8 +226,284 @@ uint64_t AieAqlQueue::AddWriteIndexAcqRel(uint64_t value) {
 }
 
 void AieAqlQueue::StoreRelaxed(hsa_signal_value_t value) {
-  atomic::Store(signal_.hardware_doorbell_ptr, uint64_t(value),
-                std::memory_order_release);
+
+  auto &driver = static_cast<XdnaDriver &>(agent_.driver());
+  std::unordered_map<void*, uint32_t> &vmem_addr_mappings = driver.GetAddrMappings();
+
+  int fd = driver.GetFd();
+  int timeout_val = driver.GetTimeout();
+
+  SubmitCmd(hw_ctx_handle_, fd, amd_queue_.hsa_queue.base_address,
+            amd_queue_.read_dispatch_id, amd_queue_.write_dispatch_id,
+            vmem_addr_mappings, timeout_val);
+}
+
+hsa_status_t AieAqlQueue::SyncBos(const std::vector<uint64_t> &bo_addrs, const std::vector<uint32_t> &bo_sizes, int fd) {
+
+  if (bo_addrs.size() != bo_sizes.size())
+    return HSA_STATUS_ERROR;
+
+  for (int i = 0; i < bo_addrs.size(); i++) {
+    clflush_data(reinterpret_cast<void *>(bo_addrs[i]), 0, bo_sizes[i]);
+  }
+
+  return HSA_STATUS_SUCCESS;
+}
+
+hsa_status_t AieAqlQueue::ExecCmdAndWait(amdxdna_drm_exec_cmd *exec_cmd,
```

Comment:
This should be in the XdnaDriver. We shouldn't be calling IOCTLs from the queue.

Reply:
Moved the work submission logic from the queue to the driver in https://github.com/ROCm/ROCR-Runtime/commit/530dfee5fea225325fff092f37dbca1f6b498e75, let me know what you think!